allennlp/training/trainer.py

```diff
@@ -145,136 +161,243 @@ def train(self) -> None:
                 if parameter.requires_grad:
                     parameter.register_hook(clip_function)
 
-        logger.info("Beginning training.")
+    def _rescale_gradients(self) -> None:
+        """
+        Performs gradient rescaling. Is a no-op if gradient rescaling is not enabled.
+        """
+        if self._grad_norm:
+            clip_grad_norm(self._model.parameters(), self._grad_norm)
+
+    def _compute_loss(self, batch: torch.Tensor, for_training: bool) -> torch.Tensor:
+        """
+        Does a forward pass on the given batch and returns the ``loss`` value in the result.
+        Also performs backpropagation if ``for_training`` is ``True``.
```

Comment:
If this is going to perform backprop, I think it should say so in its name: `_compute_and_backprop_loss`.  It's pretty unintuitive to expect a `_compute_loss` function to also do backprop.

Reply:
Actually, why put backprop in this function at all?  Why not just leave it in `_train_epoch`?  Seems cleaner.

Reply:
done